### PR TITLE
Don't propagate device APIs to parent loops

### PR DESCRIPTION
**Pull request by dsharletg**
I think that only loops that are directly involved in GPU scheduling should be explicitly marked as such, and that regular loops in the body of GPU kernels should remain "Parent" loops.

The motivation for this is that many optimizations need to respect GPU loops, and when all loops that run on a GPU are labelled GPU loops, the optimizations can't run, even though the loop has nothing to do with GPU scheduling.

### CODE REVIEW

**Comment by dsharletg** [comment]
BTW, I should point out: I ran all the tests on OpenCL and CUDA, and I didn't notice any failures due to this.

**Comment by zvookin** [comment]
I'll need to look at this carefully. At some point, I decided that there is
an invariant where after some lowering pass, the Parent designation should
be entirely removed and lowering passes should be able to rely on knowing
exactly which device the code is going to run on. That said, I would expect
the optimization passes really only care about whether the loop is one of
the magic block/thread/whatever loops, not what device is going to run on.
This raises the question of what the DeviceAPI means vs. the magic loop
variable naming thing. Are optimizations looking at the DeviceAPI? If so,
should they be?

**Comment by dsharletg** [comment]
The optimization that is giving me problems right now is RemoveTrivialForLoops, and yes, it needs to look at the device API. It can't remove loops that are relevant to the GPU loop nest, but it really should/needs to remove loops that are in a device kernel that are irrelevant to the GPU loop nest.

I don't see why it would be necessary for a lowering pass to care if a parent loop is running on the GPU or not. A loop should be a loop, whether it runs on a GPU or not. The only ones that should care are those that look at the GPU loop nest relevant loops (thread id/block id loops), and those are explicitly marked with the relevant device API, so that behavior is unchanged by this.

**Comment by abadams** [comment]
There are lots of places that explicitly check for the magic names with
substring comparisons. If this goes through they could instead check the
device_api field.

**Comment by zvookin** [comment]
The reason stuff cares is because there is an algorithm that looks for
device boundaries and having to deal with multiple values that mean the
same thing is complicated. That said, I think Default was the problem, not
Parent. However, we should make it clear when code is checking for a GPU
implicit loop (block/thread) vs. what device it is on. If we are going to
encode this in the DeviceAPI, then we should make it such that Parent and
whatever DeviceAPI parent maps to are not treated as alternate names for
the same thing. I.e. I think RemoveTrivialForLoops should probably
is is_gpu_var(loop->name) instead of looking at the DeviceAPI.

-Z-

**Comment by zvookin** [comment]
I'm all for getting rid of the reliance on the magic names, but if we do
that, I think we need more info than just which device API is used as the
loop levels (block vs. thread) are distinguished and sometimes code needs
which it is.

High-level, the reason I'm pushing back is I don't think the design is
clear. One reason having the DeviceAPI made totally concrete is useful is
that one can write a visitor for a specific GPU API and it can know to
ignore any loop that is not its own API. If we preserve Parent, then it
needs to keep a stack of which API it is in. I ok with specifying IR this
way, but I don't think it is what I intended when I started doing this.
(And it isn't clear to me it is the right thing. I.e. it seems like we're
heading toward a "loop type" concept where Parent becomes "not special" and
we add distinct values for "CUDA block loop," "CUDA threads loop," "OpenCL
block loop," "OpenCL threads loop" etc. Or perhaps we keep the two things
split into API and loop type, which seems a bit better to me.

**Comment by dsharletg** [comment]
I think you are thinking of this: https://github.com/halide/Halide/blob/master/src/InjectHostDevBufferCopies.cpp#L67

But it already handles Parent loops correctly.

RemoveTrivialForLoops actually already does look at the loop var name. However, on Hexagon (or any other general offloading processor we might support in the future), there is no explicit 'Hexagon var', because Hexagon can execute any loops it wants. The case I want to handle is people should be able to say hexagon() without specifying a loop of a func, which means that Var::outermost() is marked as DeviceAPI::Hexagon, which is the only way to identify it as a Hexagon loop.

This means that remove trivial for loops can't remove any trivial Hexagon for loops, even if they're nested inside the Hexagon code (and not just the outermost one which needs to remain as it is relevant to the schedule).

**Comment by dsharletg** [comment]
> High-level, the reason I'm pushing back is I don't think the design is
clear. One reason having the DeviceAPI made totally concrete is useful is
that one can write a visitor for a specific GPU API and it can know to
ignore any loop that is not its own API.

I don't think that's the right thing for a device specific visitor to do. Hypothetically, if there were some code with a loop of a different device API nested inside another device API, surely it's not the right thing to do to simply ignore the for loop. That either needs to be an error (e.g. launching a CUDA kernel from inside an OpenCL kernel is not valid code), or the visitor needs to handle it correctly.

**Comment by zvookin** [comment]
Sigh. We need to fix the design. The original intent of Parent was to allow
manipulating IR without knowing what device the code might run on. It is
not supposed to matter whether one uses Parent or whatever the parent
happens to be. This PR is changing the design to use the DeviceAPI to
indicate that a loop is a transition boundary to a different device. This
is sorta kinda maybe sensible if one looks at the world from a very GPU
kernel-ish viewpoint. Hexagon is more like RPC though. In this proposed
model, I would expect a threads loop to be Parent, not whatever GPU API it
runs on. Though I suppose you could conceivably say it is a transition from
blocks to threads.

To be very explicit in my objection, this places a constraint on IR that
all loops must be Parent unless the are a device transition. This is
definitely a change. If we make this change, I think we should rename at
least the device_api field and probably DeviceAPI::Parent. We definitely
need to document the invariant.

Perhaps a simpler fix is to have the hexagon scheduling primitive give a
special name to the outermost loop var and have RemoveTrivialForLoops treat
that name as a GPU var.

-Z-

**Comment by dsharletg** [comment]
> This PR is changing the design to use the DeviceAPI to
indicate that a loop is a transition boundary to a different device.

> To be very explicit in my objection, this places a constraint on IR that
all loops must be Parent unless the are a device transition.

I'm not sure this is the case. To give an example, this is what a 2D OpenCL kernel still looks like with this change:

    let device_malloc_result$2 = halide_device_malloc(input.buffer, halide_opencl_device_interface())
    assert((device_malloc_result$2 == 0), device_malloc_result$2)
    let copy_to_device_result$8 = halide_copy_to_device(input.buffer, halide_opencl_device_interface())
    assert((copy_to_device_result$8 == 0), copy_to_device_result$8)
    let copy_to_device_result$7 = halide_copy_to_device(i1.buffer, halide_opencl_device_interface())
    assert((copy_to_device_result$7 == 0), copy_to_device_result$7)
    let copy_to_device_result$6 = halide_copy_to_device(blur2.buffer, halide_opencl_device_interface())
    assert((copy_to_device_result$6 == 0), copy_to_device_result$6)
    parallel<OpenCL> (blur2.s0.y.__block_id_y, 0, ((blur2.extent.1 + 15)/16)) {
      parallel<OpenCL> (blur2.s0.x.__block_id_x, 0, ((blur2.extent.0 + 15)/16)) {
        parallel<OpenCL> (.__thread_id_y, 0, 16) {
          parallel<OpenCL> (.__thread_id_x, 0, 16) {
            let blur2.s0.y.__thread_id_y.base = min(((blur2.s0.y.__block_id_y*16) + blur2.min.1), ((blur2.min.1 + blur2.extent.1) + -16))
            let blur2.s0.x.__thread_id_x.base = min(((blur2.s0.x.__block_id_x*16) + blur2.min.0), ((blur2.min.0 + blur2.extent.0) + -16))
            allocate sum[uint16 * 1 * 1]
            produce sum {
              sum[0] = (uint16)0
            } update sum {
              for (sum.s1.i1.y$r, 0, 3) {
                for (sum.s1.i1.x$r, 0, 3) {
                  sum[0] = (sum[0] + (i1[(sum.s1.i1.x$r + (sum.s1.i1.y$r*3))]*input[((((blur2.s0.x.__thread_id_x.base + .__thread_id_x) + sum.s1.i1.x$r) - blur2.min.0.required) + ((((blur2.s0.y.__thread_id_y.base + .__thread_id_y) + sum.s1.i1.y$r) - blur2.min.1.required)*(input.x.extent_realized.s + 2)))]))
                }
              }
            }
            blur2[(((blur2.s0.x.__thread_id_x.base + .__thread_id_x) + ((blur2.s0.y.__thread_id_y.base + .__thread_id_y)*blur2.stride.1)) - (blur2.min.0 + (blur2.min.1*blur2.stride.1)))] = sum[0]
            free sum
          }
        }
      }
    }
    free input
    set_dev_dirty(blur2.buffer, (uint8)1)

Note that the OpenCL device API is doing more than just marking transitions between devices. It's saying that this loop is part of a kernel loop nest.

It seems to me that there still ought to be a plain old for loop that doesn't have any meaning beyond "this is a for loop". That's what we're currently missing, and this PR brings back with DeviceAPI::Parent (though I agree maybe it should be renamed).

**Comment by zvookin** [comment]
The proper nesting is guaranteed elsewhere. The entire point of lowering is
to make code more explicit so it is easier for later passes to deal with.
To take your example of a change in device API being an error, is going
from "Parent" to "CUDA" an error? Well it depends on whether one was
already in CUDA doesn't it? Perhaps correct IR never uses a DeviceAPI other
than Parent unless it is actually changing the API. E.g. in the Hexagon
case, going from a loop that is on Hexagon to Hexagon again doesn't result
in another RPC. (Note that in CUDA that is absolutely a valid thing to do.
Kernels are now allowed to launch other kernels. We don't support it, but
at some point we probably should. Heck, if you have umpteen Hexagon
instances, maybe having them invoke on other Hexagons is valid too.)

Basically, the proposal is to switch from device_api being little more than
a declaration that the code runs on a particular device to something else.
I'm fine with that, but we need to make the design a bit more concrete and
make some other changes around it.

-Z-

**Comment by zvookin** [comment]
Another way to go would be to introduce a new IR node to mark device API
transitions and then to kill off the device_api field. We could also
cleanup the magic for loop variable name thing by adding information to the
for loop which indicates special meaning. E.g. "this is a blocks loop" and
"this is a threads loop". We could add ones like "don't remove this loop
even if you think it is trivial". I think there was a use case for this in
Abe's vertex shader stuff as well.

-Z-

**Comment by dsharletg** [comment]
> The entire point of lowering is to make code more explicit so it is easier for later passes to deal with.

I agree... in this case, this particular lowering is making the IR harder to deal with, not easier :)

> Well it depends on whether one was already in CUDA doesn't it?

Sure, but does it actually help such a visitor to have every for loop explicitly marked? The visitor would still need to track the last for loop it cared about either way? Since it needs to store this information anyways, it seems totally reasonable for a visitor to need to check and see if a for loop is either just a for loop, or something else that it really cares about.

> (Note that in CUDA that is absolutely a valid thing to do. Kernels are now allowed to launch other kernels. We don't support it, but at some point we probably should. Heck, if you have umpteen Hexagon instances, maybe having them invoke on other Hexagons is valid too.)

I totally agree, and I think that this change actually makes this possible, not the other way around. Right now, if there are a bunch of nested kernels, all of the loops of which are labelled with a device API, that is ambiguous between "this is a nested kernel" and "this is just some random nested for loops". With this change, if the device API is set to something interesting on a nested for loop, that actually means something useful. For example, currently it could be an error to encounter a Hexagon loop within another Hexagon loop. 

> I'm fine with that, but we need to make the design a bit more concrete and
make some other changes around it.

Can you clarify what it is you want to see? I'm happy to do something else here, I just don't know what that is. The only thing I'm currently thinking is necessary is changing DeviceAPI::Parent to DeviceAPI::None or something like that.

**Comment by zvookin** [comment]
I'l let Andrew weigh in, but I think the better IR design is to introduce a
specific Stmt to mark the API transition. This would improve the Hexagon
situation by no longer requiring a trivial for loop to represent an RPC.

Separately, switching to an enum in the For node which indicates how it is
special would likely be good. Last time I looked at this there were a few
cases where a variable name was all that was passed, not the whole For, so
this might be slightly more difficult than it otherwise would be.

Adding an IR node is somewhat painful as it requires touching a fair umber
of visitors. Though in this case I think most of them get to ignore the
node and those that don't might become simpler. That said it is still not
trivial by any means and unless others agree that the improvement is
worthwhile, I won't insist we go down this path.

Baring this, I think we should rename For::device_api to
For::device_transition and rename DeviceAPI::Parent to DeviceAPI::None. I'd
like there to be some level of invariant specification as to when
For::device_transition can be something other than DeviceAPI::None.
(Ideally this would also be checked, but we're very lacking in IR
validation in general.)

-Z-

**Comment by dsharletg** [comment]
I just pushed a change to rename DeviceAPI::Parent to DeviceAPI::None.

I'm not sure that device_transition is the right name for device_api - note there are still nested loops explicitly marked with device APIs that don't indicate a transition. I think that this field really means "this loop is implemented by something other than a bit of control flow (such as a GPU kernel loop)". Whether or not such a loop is a transition is a higher level bit of info that things like InjectHostDevBufferCopies need to determine by investigating pairs of loops (and their device_apis).

Regarding the more involved changes - I agree that a bit of IR that indicates the body should be offloaded would make a lot of sense, but I know that we've generally strongly avoided adding new IR node types in the past. I'm not opposed, but I still think this PR would also be a good change.

**Comment by dsharletg** [comment]
The CI check failure seems to be a flake (travis couldn't download cmake again). I restarted it.

**Comment by abadams** [comment]
I went through all the places where we examine device_api, and 90% of them
seem to be detecting what sort of gpu API is used given that a transition
has occurred. That is, uses are mostly nested inside is_gpu_var.

So I'm in favor of only using device_api to mark the magic loops that are
logically parallel loops but are implemented as kernel launches. It would
also be nice to kill usage of is_gpu_var and replace it with some helper
method on the device_api instead (!= Host && != None?)

**Comment by dsharletg** [comment]
I agree, it would be nice to get rid of is_gpu_var. Did you want to try to make that change as part of this PR? I suppose it should be doable now.

**Comment by dsharletg** [comment]
Hmm, there are a few uses of is_gpu_var that can be easily replaced by a check of device_api, but there are others that I'm not so sure about. I'd prefer to leave them to another change. I'll make the easy ones now though.

**Comment by zvookin** [comment]
Actually, I think it would be better to just leave that for now.

**Comment by abadams** [comment]
sgtm
